### ISSUE TITLE
[Feature Request] Automatic unique task naming

### ISSUE THREAD

**Issue by mmiller-max** [allegroai/clearml#543]
It would be nice if `Task.init` and `Task.create` could have an automatic way of naming tasks that could be optionally used.

The reason for this request is that often I'll run a few tasks, use the GUI to select one or more to look at and then use the python SDK to pull those task(s) and do some further processing with them. Having to type the task ID, or finding the correct ID in a list of IDs, isn't straightforward as they are UUIDs and not easily readable, unlike (for example) 'Task001`, `Task002` etc.

I can do some logic when naming tasks to do something similar, but unfortunately it is difficult to make this thread safe, i.e. if two separate machines create a task at the same time it is difficult to ensure they are uniquely named.

I would propose adding an option to the two functions above to use default unique names which either increment a base name (obviously this needs to be done carefully) or uses more human-readable names, a bit like docker containers. (other experiment tracking software tends to do one of these two approaches). We would need to be worked out how the automatic naming interacts with more manual naming. Perhaps it's a project setting?

I'm happy to try to work on adding this as a feature, but it would be useful to know if this has been discussed already and whether or not you'd want it as a feature? Cheers

**Comment by jkhenning** [comment]
Hi @mmiller-max,

It sounds like a cool feature, albeit not without complications (as you've mentioned regarding synchronization etc.).
We'll happy to discuss possibilities in terms of approach (i.e. internal logic vs some external injected logic) and implementation if you have any suggestions, and contributions are always appreciated and welcomed 🙂 

As a first step, I was thinking we can support a callable in `task_name`, easily allowing a user-controlled naming scheme with little infrastructure. Next steps can include configurable internal logic, perhaps alongside some server-side support (such as synchronization?) - WDYT?

**Comment by mmiller-max** [comment]
@jkhenning thanks for getting back to me. 

> As a first step, I was thinking we can support a callable in `task_name`

This sounds like a good first step yep. How easy would it be to "lock" the server whilst the task is being created (e.g. to ensure two names can't be identical), is there functionality already for this or not currently? (A potential solution, for example, would be to use a Redis lock when creating a task).

Unfortunately I don't have loads of time to work on this, so it may not happen very fast, but will do what I can!

**Comment by jkhenning** [comment]
> A potential solution, for example, would be to use a Redis lock when creating a task

My thoughts exactly 🙂 - not available yet, but entirely possible to add...
In such a scenario, we won't lock the server, but simply create a new endpoint that will use a locked redis key (based on whichever criteria we decide, like project ID?) and guarantee an atomic increment, or something similar. Does that sound reasonable?

**Comment by mmiller-max** [comment]
Yep that sounds great and kind of what I'd imagined!

**Comment by bmartinn** [comment]
@mmiller-max since adding a system wide lock will probably take some time (i.e. API version bump backend support etc.)

> or uses more human-readable names, a bit like docker containers. (

This probably does not require any locks, as the naming is random, and that might be relatively easy to implement and machine safe (aka thread safe). wdyt?

**Comment by mmiller-max** [comment]
> > or uses more human-readable names, a bit like docker containers. (
> 
> This probably does not require any locks, as the naming is random, and that might be relatively easy to implement and machine safe (aka thread safe). wdyt?

I agree it's an easier route to getting something working and therefore a good place to start, but unsure how it can be fully thread safe without a lock. I agree that because it's random and not incrementing you're much less likely to get two identical tasks names at the same time, but there is still a small chance. Still, that is probably acceptable as a halfway step to the system wide lock approach. Do you agree, or do you think the lock approach is a bit too big to implement?
